toc/rfc/rfc-draft-v2-app-manifests.md

```diff
@@ -0,0 +1,613 @@
+final._
+
+#### Droplets
+
+Similar to the `--droplet` flag on `cf push` and the `package.path` node for
+app resources proposed above, there could be a way to upload droplets when
+pushing with a manifest.
+
+Example manifest:
+```yaml
+---
+apps:
+  - name: app1
+    droplet:
+      path: /local/path/to/droplet.tgz
+```
+
+#### Remote Files
+
+In addition to local paths, the manifest could be extended to support remote
+files for packages and droplets. Unlike local file paths, these could be
+handled server-side, which would further enable GitOps workflows. However,
+handling remote files would introduce some of the same access and autentication
+issues faced by docker-lifecycle-app's remote image registries.
+
+Example manifest:
+```yaml
+---
+apps:
+  - name: app1
+    package:
+      path: https://example.com/app_source.tgz
+      access_credentials:
+        - type: basic
+          username: new-user
+          password: new-pass
+        - type: basic
+          username: old-user
+          password: old-pass
+```
+
+#### Buildpack Fallback URL
+
+Buildpacks can currently be specified as either admin-installed buildpack names
+or remote urls. Since the set of installed buildpacks can vary between Cloud
+Foundry deployments, it is difficult for manifest authors to know whether they
+should reference a buildpack by name, or if they need to provide a url.
+
+As requested in [this
+issue](https://github.com/cloudfoundry/cloud_controller_ng/issues/1358),
+manifests could allow referencing a buildpack by name, while also specifying a
+url that will be used if the referenced buildpack is not available.
+
+Example manifest:
+```yaml
+---
+apps:
+  - name: app1
+    lifecycle:
+      type: buildpack
+      buildpacks:
+        - name: my_buildpack
+          url: git.example.com/my_buildpack
+```
+
+#### Shared Resources
+
+Some resources can be shared across multiple spaces (currently, routes and
+service instances). Sharing resources is not currently possible via app
+manifests. Since these resources are no longer fully-contained within a space,
+they require configuration split across multiple space-level manifests.
+
+One option could be to add a boolean `shared` node to sharable resources, so
+they are not automatically created in the destination space, but can still be
+bound/mapped/etc to.
+
+Source manifest:
+```yaml
+---
+routes:
+  - url: route.example.com
+    destinations:
+      - app:
+          name: app1
+service_instances:
+  - name: my-service-instance
+    bindings:
+      apps:
+        - name: app1
+    share_to:
+      - org: my-org
+        space: my-space
+```
+
+Destination manifest:
+```yaml
+---
+routes:
+  - url: route.example.com
+    shared: true
+    destinations:
+      - app:
+          name: app1
+service_instances:
+  - name: my-service-instance
+    shared: true
+    bindings:
+      apps:
+        - name: app1
+```
+
+#### Network Policies
```

Comment:
I know network policies is listed under the "future" section, but I want to elaborate here why these were never added to app manifests. tl;dr:  *Permissions issues.**.

Network policies often bridge spaces. A frontend app may be in one space while a backend app may be in another. In order to create a network policy you have to have certain permissions in both of those spaces. 
* User-A has permissions in both spaces space-frontend and space-backend.
* User-B has permissions only in space-backend.
* User-A pushes this app. Great she has permissions to create network policies!
* User-B pushes the app... what happens? Does it fail because user-b doesn't have permissions? Does the app get pushed, but WITHOUT the network policies? Does it succeed because the network policies aren't changing? 

I would love for network policies to be gitops-able. I think we can figure out a reasonable solution that -- while maybe not intuitive -- would at least be written down and deterministic. But to me the solution is not an obvious one.

Reply:
Related to network policy permissions. We need to use the user's token to call the network policy server to be able to take these actions (unless Cloud Controller wants to duplicate that logic and check scopes / permissions itself). Currently most of the manifest processing happens in a [delayed job](https://github.com/cloudfoundry/cloud_controller_ng/blob/main/app/controllers/v3/space_manifests_controller.rb#L39) that wouldn't have access to the user's token (and shouldn't). We might need to make those calls earlier after parsing the manifest, but before enqueuing the job.

Reply:
In my mind, both network policies and shared {service instances, routes} will behave the same as they would with imperative CLI commands. CC will attempt to perform the necessary operations, using the user's token (like Tim is saying above). If the user doesn't have sufficient permissions to complete the operation, then creating the policy will fail, and  thus applying the manifest will also fail.

So applying this manifest:

```yaml
---
network_policies:
  - destination:
      app:
        name: my-destination-app
    source:
      app:
        name: source-app
```
will behave the same as executing this command:
```sh
cf add-network-policy my-source-app --destination-app=my-destination-app
```